src/common/tests/ad_mem_tests.c

```diff
@@ -840,7 +841,7 @@ adt_teardown(void **state)
 	int	rc;
 
 	printf("close ad_blob\n");
-	rc = ad_blob_close(adt_bh);
+	rc = ad_blob_destroy(adt_bh);
```

Comment:
There are still a reference leak that blob_fini() is not called properly.

In ad_blob_prep_open() if it is dummy_blob, ref will be 2:
                      blob = dummy_blob;
 469                         D_DEBUG(DB_TRACE, "found dummy blob, refcount=%d\n", blob->bb_ref);
 470                         blob_addref(blob);

maybe 
blob_decref(adt_bh.bh_blob)
need be called before destroy?

Reply:
To be clear, the global variable "dummy_blob" has no refcount (it is reset to NULL when the last opener closed), multiple openers can find the same "dummy_blob" on open, instead of creating their own instance. 

The main reason dummy_blob has no refcount is because I don't want to pin it in memory so I can test blob_open(). If it has refcount, then the global blob is never going to be freed, so blob_open() will not be tested at all.

Reply:
I added some codes it showed there are still leaks somewhere:
[root@daos1 daos]# git diff .
diff --git a/src/common/ad_mem.c b/src/common/ad_mem.c
index d91a228551..bc46f35d12 100644
--- a/src/common/ad_mem.c
+++ b/src/common/ad_mem.c
@@ -562,6 +562,7 @@ ad_blob_destroy(struct ad_blob_handle bh)

        /* TODO: remove the file */
        blob_close(blob);
+       printf("(blob->bb_ref: %d\n", blob->bb_ref);
        blob_decref(blob);
        return 0;
 }
close ad_blob
(blob->bb_ref: 2
[==========] 16 test(s) run.

Reply:
thanks, I will check

Reply:
found the issue, it's an issue in test program, I forget to publish/cancel the last reserve in adt_delayed_free_1(), because it's not a bug in allocator itself, so I will submit another patch to fix it.